src/windows/keycodes.rs

```diff
@@ -1,328 +1,138 @@
 use crate::rdev::Key;
+use std::convert::TryInto;
+use winapi::shared::minwindef::WORD;
 
-/// https://docs.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes
-/// We redefined here for Letter and number keys which are not in winapi crate (and don't have a name either in win32)
-const ALT: u16 = 164;
-const ALT_GR: u16 = 165;
-const BACKSPACE: u16 = 0x08;
-const CAPS_LOCK: u16 = 20;
-const CONTROL_LEFT: u16 = 162;
-const CONTROL_RIGHT: u16 = 163;
-const DELETE: u16 = 46;
-const DOWN_ARROW: u16 = 40;
-const END: u16 = 35;
-const ESCAPE: u16 = 27;
-const F1: u16 = 112;
-const F10: u16 = 121;
-const F11: u16 = 122;
-const F12: u16 = 123;
-const F2: u16 = 113;
-const F3: u16 = 114;
-const F4: u16 = 115;
-const F5: u16 = 116;
-const F6: u16 = 117;
-const F7: u16 = 118;
-const F8: u16 = 119;
-const F9: u16 = 120;
-const HOME: u16 = 36;
-const LEFT_ARROW: u16 = 37;
-const META_LEFT: u16 = 91;
-const PAGE_DOWN: u16 = 34;
-const PAGE_UP: u16 = 33;
-const RETURN: u16 = 0x0D;
-const RIGHT_ARROW: u16 = 39;
-const SHIFT_LEFT: u16 = 160;
-const SHIFT_RIGHT: u16 = 161;
-const SPACE: u16 = 32;
-const TAB: u16 = 0x09;
-const UP_ARROW: u16 = 38;
-const PRINT_SCREEN: u16 = 44;
-const SCROLL_LOCK: u16 = 145;
-const PAUSE: u16 = 19;
-const NUM_LOCK: u16 = 144;
-const BACK_QUOTE: u16 = 192;
-const NUM1: u16 = 49;
-const NUM2: u16 = 50;
-const NUM3: u16 = 51;
-const NUM4: u16 = 52;
-const NUM5: u16 = 53;
-const NUM6: u16 = 54;
-const NUM7: u16 = 55;
-const NUM8: u16 = 56;
-const NUM9: u16 = 57;
-const NUM0: u16 = 48;
-const MINUS: u16 = 189;
-const EQUAL: u16 = 187;
-const KEY_Q: u16 = 81;
-const KEY_W: u16 = 87;
-const KEY_E: u16 = 69;
-const KEY_R: u16 = 82;
-const KEY_T: u16 = 84;
-const KEY_Y: u16 = 89;
-const KEY_U: u16 = 85;
-const KEY_I: u16 = 73;
-const KEY_O: u16 = 79;
-const KEY_P: u16 = 80;
-const LEFT_BRACKET: u16 = 219;
-const RIGHT_BRACKET: u16 = 221;
-const KEY_A: u16 = 65;
-const KEY_S: u16 = 83;
-const KEY_D: u16 = 68;
-const KEY_F: u16 = 70;
-const KEY_G: u16 = 71;
-const KEY_H: u16 = 72;
-const KEY_J: u16 = 74;
-const KEY_K: u16 = 75;
-const KEY_L: u16 = 76;
-const SEMI_COLON: u16 = 186;
-const QUOTE: u16 = 222;
-const BACK_SLASH: u16 = 220;
-const INTL_BACKSLASH: u16 = 226;
-const KEY_Z: u16 = 90;
-const KEY_X: u16 = 88;
-const KEY_C: u16 = 67;
-const KEY_V: u16 = 86;
-const KEY_B: u16 = 66;
-const KEY_N: u16 = 78;
-const KEY_M: u16 = 77;
-const COMMA: u16 = 188;
-const DOT: u16 = 190;
-const SLASH: u16 = 191;
-const INSERT: u16 = 45;
-// const KP_RETURN: u16 = 13;
-const KP_MINUS: u16 = 109;
-const KP_PLUS: u16 = 107;
-const KP_MULTIPLY: u16 = 106;
-const KP_DIVIDE: u16 = 111;
-const KP0: u16 = 96;
-const KP1: u16 = 97;
-const KP2: u16 = 98;
-const KP3: u16 = 99;
-const KP4: u16 = 100;
-const KP5: u16 = 101;
-const KP6: u16 = 102;
-const KP7: u16 = 103;
-const KP8: u16 = 104;
-const KP9: u16 = 105;
-const KP_DELETE: u16 = 110;
+macro_rules! decl_keycodes {
```

Comment:
@Narsil how do you feel about the decl_keycode macro?

Reply:
Seems much better. I'm usually against magic, but it's always good when it prevents us from shooting in the foot.